mpisppy/convergers/norms_and_residuals.py

```diff
@@ -0,0 +1,156 @@
+
+    PHB.comms["ROOT"].Allreduce(local_sum, global_sum, op=MPI.SUM)
+    return np.sqrt(global_sum[0])
+
+
+
+
+def primal_residuals_norm(PHB):
+    """
+    Compute the scaled primal residuals Euclidean norm.
+    """
+    local_sum = np.zeros(1)
+    global_sum = np.zeros(1)
+    for s in PHB.local_scenarios.values():
+        for node in s._mpisppy_node_list:
+            ndn = node.name
+            nlen = s._mpisppy_data.nlens[ndn]
+            xbars = np.fromiter((s._mpisppy_model.xbars[ndn,i]._value
+                                  for i in range(nlen)), dtype='d')
+            nonants_array = np.fromiter((v._value for v in node.nonant_vardata_list),
+                                        dtype='d', count=nlen)
+            resid = np.abs(nonants_array - xbars)
+            if not s._mpisppy_data.has_variable_probability:
+                local_sum[0] += s._mpisppy_data.prob_coeff[ndn] * np.sum(np.multiply(resid, resid))
+            else:
+                # rarely-used overwrite in the event of variable probability
+                prob_array = np.fromiter((s._mpisppy_data.prob_coeff[ndn_i[0]][ndn_i[1]]
+                                          for ndn_i in s._mpisppy_data.nonant_indices
+                                          if ndn_i[0] == ndn),
+                                         dtype='d', count=nlen)
+                local_sum[0] += np.dot(prob_array, resid)
+    PHB.comms["ROOT"].Allreduce(local_sum, global_sum, op=MPI.SUM)
+    return np.sqrt(global_sum[0])
+
+
+
+def dual_residuals_norm(PHB, xbar_cache, curr_iter):
+    """
+    Compute the scaled primal residuals Euclidean norm.
+    """
+    local_sum = np.zeros(1)
+    global_sum = np.zeros(1)
+    for sname, s in PHB.local_scenarios.items():
+        current_xbar = np.array(xbar_cache[curr_iter][sname])
+        prev_xbar = np.array(xbar_cache[curr_iter-1][sname])
+        for node in s._mpisppy_node_list:
+            ndn = node.name
+            nlen = s._mpisppy_data.nlens[ndn]
+            rhos = np.fromiter((s._mpisppy_model.rho[(ndn, i)]._value for i in range(nlen)),
+                               dtype='d', count=nlen)
+            resid = np.multiply(rhos, np.abs(current_xbar - prev_xbar))
+            if not s._mpisppy_data.has_variable_probability:
+                local_sum[0] += s._mpisppy_data.prob_coeff[ndn] * np.sum(np.multiply(resid, resid))
+            else:
+                # rarely-used overwrite in the event of variable probability
+                prob_array = np.fromiter((s._mpisppy_data.prob_coeff[ndn_i[0]][ndn_i[1]]
+                                          for ndn_i in s._mpisppy_data.nonant_indices
+                                          if ndn_i[0] == ndn),
+                                         dtype='d', count=nlen)
+                local_sum[0] += np.dot(prob_array, resid)
+    PHB.comms["ROOT"].Allreduce(local_sum, global_sum, op=MPI.SUM)
+    return np.sqrt(global_sum[0])
```

Comment:
These are largely similar to the functions above. I think they should be combined with a flag for scaling or not.

Reply:
These functions could/should be factored, but it not quite as simple as a flag for scaling, so I am not going to do it right now.